Add CO2 sensing with SCD4X

Coloradohusky opened the pull request:
This PR adds support for CO2 sensing with the SCD40/SCD41 sensors. These sensors also support temperature and relative humidity.

Currently, it is using `startLowPowerPeriodicMeasurement`, which has an update interval of ~30 seconds. However, the datasheet mentions a `measure_single_shot` for lower power use, but says it is limited to SCD41 sensors only. Nonetheless, I tried it on my SCD40 sensor and it appeared to work fine. `measure_single_shot` also includes a delay of 5 seconds for each use.

Average Supply Current:
- High Performance Mode: 11 mA
- Low Power Mode: 2.8 mA
- Idle Single Shot Mode: 1.1 mA for 1 min and 0.36 mA for 5 min interval
- Power-Cycled Single Shot Mode: 0.18 mA for 10 min, 0.09 mA for 20 min, 0.03 mA for 1 hour interval

Other than that, the only differences between SCD40 and SCD41 are a wider CO2 interval and higher accuracy.

See [SCD4x Low Power Operation](https://sensirion.com/media/documents/077BC86F/62BF01B9/CD_AN_SCD4x_Low_Power_Operation_D1.pdf) and the [Datasheet](https://sensirion.com/media/documents/48C4B7FB/64C134E7/Sensirion_SCD4x_Datasheet.pdf) for more information.

Thanks

Coloradohusky committed:
add CO2 sensing with SCD4X

caveman99 committed:
Merge branch 'master' into scd4x

caveman99 commented on src/mesh/generated/meshtastic/telemetry.pb.h:
```diff
@@ -71,7 +71,9 @@ typedef enum _meshtastic_TelemetrySensorType {
-    meshtastic_TelemetrySensorType_CUSTOM_SENSOR = 29
+    meshtastic_TelemetrySensorType_CUSTOM_SENSOR = 29,
+    /* SCD40/SCD41 CO2 sensor */
+    meshtastic_TelemetrySensorType_SCD4X = 30
```
These files are generated and will be overwritten with the newxt commit. You need to add this to the protobuf repository instead.

Coloradohusky replied:
That would make sense - created pull request https://github.com/meshtastic/protobufs/pull/572

fifieldt commented on src/detect/ScanI2CTwoWire.cpp:
```diff
@@ -378,6 +378,11 @@ void ScanI2CTwoWire::scanPort(I2CPort port, uint8_t *address, uint8_t asize)
                 SCAN_SIMPLE_CASE(MLX90632_ADDR, MLX90632, "MLX90632 IR temp sensor found\n");
                 SCAN_SIMPLE_CASE(NAU7802_ADDR, NAU7802, "NAU7802 based scale found\n");
 
+            case SCD4X_ADDR:
```
if you're doing another update, could use the SCAN_SIMPLE_CASE macro to keep in line with other entries here

fifieldt commented:
Since this one has temperature and humidity in addition to co2, it looks like it spans our EnvironmentMetrics and AirQualityMetrics. Do you have a sense for how easy it would be to put the co2 in AirQualityMetrics and keep the temp/humidity in EnvironmentMetrics?

Coloradohusky committed:
Merge remote-tracking branch 'upstream/master' into scd4x

Coloradohusky committed:
update from upstream

Coloradohusky commented:
I'll work on moving it over to AirQualityMetrics instead of EnvironmentMetrics

Coloradohusky committed:
Merge remote-tracking branch 'upstream/master' into scd4x

Coloradohusky committed:
Merge remote-tracking branch 'upstream/master' into scd4x

Coloradohusky commented:
Hm yeah, I'm looking closer at AirQualityModule and it's going to be more involved than I'd hoped - AirQualityModule explicitly wants a Adafruit_PM2.5 sensor (`aqi = Adafruit_PM25AQI()`) so I'm not entirely sure how to handle that with the addition of the CO2 metric

fifieldt commented:
Thanks for the heads up! Will take a look in the next few days.

fifieldt committed:
Merge branch 'master' into scd4x

fifieldt committed:
Abstract AirQualityTelemetry module

Previously the module was designed around a single sensor.
This patch brings it into line with the same design as
EnvironmentTelemetry.

fifieldt commented:
@Coloradohusky - I hacked together some things over at 
https://github.com/fifieldt/meshtastic-firmware/tree/scd4x

It compiles ... but I haven't got it further than that :)

Do you think you'd be able to hack from there and test with your sensor?

fifieldt committed:
Re-add I2C scan, fix frame.

fifieldt commented:
@Coloradohusky , have hacked this a bit more. In theory you might also get a screen on your UI with CO2 readings too. Can you test & improve? =D

fifieldt committed:
Merge branch 'master' into scd4x

Coloradohusky commented:
Was checking it out just now! Will keep testing 😄

fifieldt commented:
OK, over to you -- I'll hold off making any more commits

thebentern committed:
Merge branch 'master' into scd4x

Coloradohusky commented:
Haven't had any luck yet - it properly detects the sensor, but doesn't grab any measurements. Not sure what's preventing the measurements but will keep testing

caveman99 committed:
Merge branch 'master' into scd4x

fifieldt committed:
Merge branch 'master' into scd4x

fifieldt commented:
A fix went in at https://github.com/meshtastic/firmware/pull/5584 that seems relevant.

b8b8 commented:
any update on this sensor recently?
![image](https://github.com/user-attachments/assets/0934b7f5-fe20-44d4-a7fa-173822c2cb5f)

caveman99 committed:
Merge branch 'master' into scd4x

caveman99 reviewed (changes requested):
@Coloradohusky i had to resolve a few merge conflicts. can you please verify this is still your intended code?

caveman99 committed:
Merge branch 'master' into scd4x

caveman99 committed:
Wupps

caveman99 committed:
Merge branch 'master' into scd4x

hafu commented on platformio.ini:
```diff
@@ -183,7 +184,7 @@ lib_deps =
 	# renovate: datasource=custom.pio depName=INA226 packageName=robtillaart/library/INA226
 	robtillaart/INA226@0.6.4
-
+	https://github.com/gjelsoe/STK8xxx-Accelerometer/archive/v0.1.1.zip
```
Why is this added?

hafu commented on src/detect/ScanI2C.cpp:
```diff
@@ -41,6 +41,12 @@ ScanI2C::FoundDevice ScanI2C::firstAccelerometer() const
+    ScanI2C::DeviceType types[] = {PMSA0031, SCD4X};
+    return firstOfOrNONE(2, types);
+}
+
```
I understand the intention. The PMSA003I measures the particles and the SCD4x CO2. When connecting both, only the values of the PMSA003I are measured and the SCD4x is ignored. Maybe differ between particles and co2, so that particles and co2 can be measured. Such an approach could be used if for example an SCD30 and SCD4x is connected.

fifieldt replied:
Generally meshtastic only supports one sensor per type at a time.

hafu commented on src/modules/Telemetry/Sensor/SCD4XSensor.cpp:
```diff
@@ -0,0 +1,52 @@
+        measurement->variant.environment_metrics.has_relative_humidity = true;
+        measurement->variant.air_quality_metrics.has_co2 = true;
+        measurement->variant.environment_metrics.temperature = temperature;
+        measurement->variant.environment_metrics.relative_humidity = humidity;
```
I tried a similar approach with an SCD30 but setting  the `environment_metrics` resulted in corrupted data for particles (non zero values). Maybe there is some memory shared along `air_quality_metrics` and `environment_metrics`?

I suggest to set the `environment_metrics` in `EnvironmentTelemetry.cpp`.

hafu reviewed (commented):
I tried to use this branch and wanted to add support for the SCD30 (#4381), but had no success. I have some questions to the changes.

I started over and tried to add the SCD30. I've got a working branch: https://github.com/hafu/meshtastic-firmware/tree/add-scd30 Adding the SCD4x should be possible too. I could add the SCD4x to my branch, but do not have a SCD4x to test.

fifieldt commented:
Thanks for the review @hafu ! Help taking the concept forward would be most welcome

fifieldt commented:
I finally acquired an SCD40. Need to solder the pins on, but after that should be able to help test this.

fifieldt commented:
I am having a saga acquiring a soldering iron :(

oscgonfer commented:
Hi! I had started doing the same rework to set it all up for our Smartcitizen hackathon next week, but probably best to start working on this on this same branch. Which commit would you recommend starting with? (There have been quite some merges.. )

Reading the comments above I'd suggest approaching the temperature and humidity of these sensors (CO2 or other multi-sensor modules) as "temperature of the co2 sensor", and add it as `co2_sensor_temperature` (or something like that?) in the protobuf. It would be good to keep these values separate from environment telemetry, because its otherwise difficult to get a good temperature reading (which is not affected by the co2 sensor operation itself).

Note: I have some of these sensors I can use to test as well (except PMSA003I). If you are at discord, feel free to ping at `oscgonfer`

hafu commented:
As mentioned before, I didn't got this branch in a working state and started over. There are also many changes in the firmware ongoing and the PMSA003I got also [some changes](https://github.com/meshtastic/firmware/pull/6569), which must be applied here.

I suggest to start a new branch with the SCD4x sensor and keep it simple. Get it working to read out the values and send them via telemetry. In my opinion the support for displays can be added in another PR, but it is a nice to have.

> Reading the comments above I'd suggest approaching the temperature and humidity of these sensors (CO2 or other multi-sensor modules) as "temperature of the co2 sensor", and add it as `co2_sensor_temperature` (or something like that?) in the protobuf. It would be good to keep these values separate from environment telemetry, because its otherwise difficult to get a good temperature reading (which is not affected by the co2 sensor operation itself).

I did another approach, at least for the SCD30: The SCD30 comes with an extra temperature/humidity sensor (RH/T something) integrated for calibration. I have to read the datasheet for the SCD4x if it is alike. Therefore I suggest to use the values for from the temperature/humidity sensor as environment metrics and calibration, if no other sensor is available.

I also wanted to start on a new branch this week, since I have now a SCD4x.

fifieldt committed:
Merge branch 'master' into scd4x

oscgonfer commented:
> > Haven't had much luck yet - it properly detects the sensor, but doesn't grab any measurements. Not sure what's preventing the measurements but will keep testing
> 
> A comment to an old comment: I had some issues with the SCD30 too. I figured out, that the SCD30 only works reliable with 5V. Maybe the 3.3V power rail is a bit to low and prevents the sensor from working properly.

I think this should be no issue... we have used SCD30 for very looong data collections (+2 years). (some SCD30 docs here https://docs.smartcitizen.me/knowledge/air/co2/Sensirion_SCD30/)

oscgonfer commented:
> > Hi @hafu,
> > We are doing a [hackathon](https://www.hackster.io/news/fab-lab-barcelona-seeed-studio-hackster-meshtastic-to-collaborate-on-next-gen-smart-citizen-kit-aa63c981164a) next week to integrate a whole bunch of air quality sensors into meshtastic (we are coming from a project called smartcitizen with a [lot of sensors supported including these](https://github.com/fablabbcn/smartcitizen-kit-2x)). For the hackathon, we have a selection of sensors that overlap with: scd30, scd40, sfa30, sen5x and sen6x series which we are mapping here: [fablabbcn#4](https://github.com/fablabbcn/smartcitizen-meshtastic/issues/4), and most of them are multisensor nodes that have to be agreed on how the firmware treats the readings.
> > I think it'd be great if we manage to all sync on this effort, and have a joint approach for integrating these... We can probably discuss the reasoning on discord or a joint call together? I think we can find an agreement on the approach, and potentially even work on a new branch together (maybe on the [hackathon repo?](https//github.com/fablabbcn/smartcitizen-meshtastic)
> > What do you think?
> 
> Sounds great! This would also be my first major contribution to the project. I did some tinkering with modules, sensors and build my customized meshtastic firmware. Getting CO2 sensors running with meshtastic is on my to do list since the beginning of the year.
> 
> I give discord another try, if this is the preferred way. Tomorrow I have some time to discuss. A call is also possible tomorrow.

I think one very good way to start would be to have a clean branch, updated with master, that has the decouple of `AirQualityTelemetry` and the PMA003I. I can give it a go, but I have no PMA sensor with me (although I have all the other sensirion ones available).

fifieldt commented:
Trying to clean this up a little for you so you can have a more pleasant hackathon.

fifieldt committed:
Merge branch 'master' into scd4x

fifieldt committed:
Fix merge

fifieldt committed:
Merge branch 'master' into scd4x

fifieldt committed:
Update main.cpp

fifieldt committed:
Merge branch 'master' into scd4x

hafu commented:
> I give discord another try, if this is the preferred way. Tomorrow I have some time to discuss. A call is also possible tomorrow.

@oscgonfer  Sorry, I got sick, but I got some energy today and tried out your PR and gave some feedback.

caveman99 committed:
Merge branch 'master' into scd4x

vidplace7 committed:
Merge branch 'master' into scd4x
